python3_user_features.rst

```diff
@@ -0,0 +1,311 @@
+    1, 2
+
+And similarly, the ``end`` argument can be used to customize the end of the line -
+this defaults to ``\n``, which is a carriage return (or *newline*):
+
+    >>> print("hello"); print("world")
+    hello
+    world
+    >>> print("hello", end=' '); print("world")
+    hello world
+
+In the above example, we had to put the print statements on the same line,
+because in interactive Python, you will be returned to the Python prompt after
+the line is executed. However, in scripts, you can do::
+
+    print("hello ", end=' ')
+    print("world")
+
+Finally, a last useful feature is that it is possible to send the output of the
+print calls to file-like objects instead of the main terminal output (the
+*standard output*):
+
+    >>> f = open('data.txt', 'w')
+    >>> print(a, b, file=f)
+    >>> f.close()
+
+or better, if you are familiar with the context manager notation:
+
+    >>> with open('data.txt', 'w') as f:
+    ...     print(a, b, file=f)
+
+Advanced unpacking
+------------------
+
+In Python 2, you can use implicit unpacking of variables to go from a list,
+tuple, or more generally any *iterable* to separate variables:
+
+    >>> a, b, c = range(3)
+    >>> a
+    0
+    >>> b
+    1
+    >>> c
+    2
+
+The number of items in the iterable on the right has to match exactly the number
+of variables on the left. However, there are cases where one might only be
+interested in the first few items of the iterable. For example, if you have a
+list of 5 items::
+
+    >>> values = range(10)
+
+and are only interested in the first two, in Python 2 you would need to do
+either:
+
+    >>> a, b, _, _, _ = values
+
+or
+
+    >>> a = values[0]
+    >>> b = values[1]
+
+Python 3 now allows users to use the ``*variable`` syntax (similar to ``*args``
+in function arguments) to avoid having to write out as many variables than items
+in the iterable::
+
+    >>> a, b, *rest = values
+    >>> a
+    0
+    >>> b
+    1
+    >>> rest
+    [2, 3, 4]
+
+The ``*`` syntax can also be used for e.g. the first variable and variables in the middle::
+
+    >>> a, *rest, b = range(5)
+    >>> a, b
+    (0, 4)
+    >>> *rest, a, b = range(5)
+    >>> a, b
+    (3, 4)
+
+This can be used for example to access the first two lines and the last line
+in a file:
+
+    >>> f = open('data.txt')
+    >>> first, second, *rest, last = f.readlines()
+    >>> f.close()
+
+Function annotations
+--------------------
+
+Since Python 3.5, it is possible to use the following syntax to annotate
+functions, to provide information on inputs/outputs. For example, it is possible
+to specify *type* annotations:
+
+    >>> def remove_spaces(x: str) -> str:
+    ...     return x.replace(' ', '')
+
+This syntax means that the input as well as the output should be a string. Now
+it turns out that Python doesn't do anything with these type annotations (there
+are still reasons why developers might want to do this, but this is not
+necessarily critical for the typical user).
```

Comment:
Note here mypy for static checking of type correctness, and that several IDEs (such as PyCharm) are already using these annotations to help static analysis of your code.